examples/binding-scala/js/src/main/scala/com/thoughtworks/todo/Main.scala

```diff
@@ -0,0 +1,143 @@
+      <h1>todos</h1>
+      <input class="new-todo" autofocus={true} placeholder="What needs to be done?" onkeydown={keyDownHandler}/>
+    </header>
+  }
+
+  @dom def todoListItem(todo: Todo): Binding[Node] = {
+    // onblur is not only triggered by user interaction, but also triggered by programmatic DOM changes.
+    // In order to suppress this behavior, we have to replace the onblur event listener to a dummy handler before programmatic DOM changes.
+    val suppressOnBlur = Var(false)
+    def ignoreEvent = { _: Event => }
+    def submit = { event: Event =>
+      suppressOnBlur := true
+      editingTodo := None
+      event.currentTarget.asInstanceOf[HTMLInputElement].value.trim match {
+        case "" =>
+          allTodos.get.remove(allTodos.get.indexOf(todo))
+        case trimmedTitle =>
+          allTodos.get(allTodos.get.indexOf(todo)) = new Todo(trimmedTitle, todo.completed)
+      }
+    }
+    def keyDownHandler = { event: KeyboardEvent =>
+      event.keyCode match {
+        case KeyCode.Escape =>
+          suppressOnBlur := true
+          editingTodo := None
+        case KeyCode.Enter =>
+          submit(event)
+        case _ =>
+      }
+    }
+    val edit = <input class="edit" value={ todo.title } onblur={if (suppressOnBlur.bind) ignoreEvent else submit} onkeydown={keyDownHandler} />;
+    <li class={s"${if (todo.completed) "completed" else ""} ${if (editingTodo.bind.contains(todo)) "editing" else ""}"}>
+      <div class="view">
+        <input class="toggle" type="checkbox" checked={todo.completed} onclick={_: Event =>
+          allTodos.get(allTodos.get.indexOf(todo)) = new Todo(todo.title, dom.currentTarget[HTMLInputElement].checked)
+        }/>
+        <label ondblclick={ _: Event => editingTodo := Some(todo); edit.focus() }>{ todo.title }</label>
+        <button class="destroy" onclick={_: Event => allTodos.get.remove(allTodos.get.indexOf(todo))}></button>
+      </div>
+      { edit }
+    </li>
+  }
+
+  @dom def mainSection: Binding[Node] = <section class="main" style:display={if (allTodos.length.bind == 0) "none" else ""}>
+    <input type="checkbox" class="toggle-all" checked={active.items.length.bind == 0} onclick={_: Event =>
+      for ((todo, i) <- allTodos.get.zipWithIndex) {
+        if (todo.completed != dom.currentTarget[HTMLInputElement].checked) {
+          allTodos.get(i) = new Todo(todo.title, dom.currentTarget[HTMLInputElement].checked)
+        }
+      }
+    }/>
+    <label for="toggle-all">Mark all as complete</label>
+    <ul class="todo-list">{ for { todo <- currentTodoList.bind.items } yield todoListItem(todo).bind }</ul>
+  </section>
+
+  @dom def filterListItem(todoList: TodoList): Binding[Node] = <li>
+    <a href={ todoList.hash } class={ if (todoList == currentTodoList.bind) "selected" else "" }>{ todoList.text }</a>
+  </li>
+
+  @dom def footer: Binding[Node] = <footer class="footer" style:display={if (allTodos.length.bind == 0) "none" else ""}>
```

Comment:
I still feel that `edit`, `mainSection` and `footer` are walls of text, with a lot of different syntactic constructs intermingled. How about we give it some whitespace treatment and extract some values here too?
Something like this?

```scala
  @dom
  def footer: Binding[Node] = {

    val onClear: Event ⇒ Seq[Todo] =
      (_: Event) => {
        val completedTodos: Seq[Todo] =
          for {
            todo <- allTodos.get if todo.completed
          } yield todo

        allTodos.get --= completedTodos
      }

    val numActive: Binding[Int] =
      active.items.length

    val filters: BindingSeq[Node] =
      for {
        todoList <- Constants(todoLists: _*)
      } yield filterListItem(todoList).bind

    <footer class="footer"
            style:display={if (allTodos.length.bind == 0) "none" else ""}>

      <span class="todo-count">
        <strong>{ numActive.bind.toString }</strong>
        {if (numActive.bind == 1) "item" else "items"} 
        left
      </span>

      <ul class="filters">{ filters }</ul>

      <button class="clear-completed"
              style:visibility={if (completed.items.length.bind == 0) "hidden" else "visible"}
              onclick={onClear}>
        Clear completed
      </button>
    </footer>
  }
```

I get that that it is an awesome property to have the source be short, but i think readability suffers for it. 
WDYT?